exporter/loadbalancingexporter/resolver_k8s.go

```diff
@@ -0,0 +1,301 @@
+	}
+	var initErr error
+	r.once.Do(func() {
+		if r.epsListWatcher != nil {
+			r.logger.Debug("creating and starting endpoints informer")
+			epsInformer := cache.NewSharedInformer(r.epsListWatcher, &corev1.Endpoints{}, 0)
+			if _, err := epsInformer.AddEventHandler(r.handler); err != nil {
+				r.logger.Error("error adding event handler to endpoints informer", zap.Error(err))
+			}
+			go epsInformer.Run(r.stopCh)
+			if !cache.WaitForCacheSync(r.stopCh, epsInformer.HasSynced) {
+				r.logger.Error("error endpoints informer not sync")
+				initErr = errors.New("endpoints informer not sync")
+				return
+			}
+		}
+	})
+	if initErr != nil {
+		return initErr
+	}
+
+	r.logger.Debug("K8s service resolver started",
+		zap.String("service", r.service),
+		zap.String("namespace", r.namespace),
+		zap.Int32s("ports", r.port))
+	return nil
+}
+
+func (r *k8sResolver) shutdown(ctx context.Context) error {
+	r.changeCallbackLock.Lock()
+	r.onChangeCallbacks = nil
+	r.changeCallbackLock.Unlock()
+
+	close(r.stopCh)
+	r.shutdownWg.Wait()
+	return nil
+}
+func newInClusterClient() (kubernetes.Interface, error) {
+	cfg, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	return kubernetes.NewForConfig(cfg)
+}
+
+func (r *k8sResolver) resolve(ctx context.Context) ([]string, error) {
+	r.shutdownWg.Add(1)
+	defer r.shutdownWg.Done()
+
+	var backends []string
+	r.endpointsStore.Range(func(address, value any) bool {
+		addr := address.(string)
+		var backend string
+		ip, err := net.ResolveIPAddr("ip", addr)
+		if err != nil {
+			stats.RecordWithTags(ctx, k8sResolverSuccessFalseMutators, mNumResolutions.M(1))
+			r.logger.Error("resolve address error", zap.String("address", addr), zap.Error(err))
+			return true
+		}
+		stats.RecordWithTags(ctx, k8sResolverSuccessTrueMutators, mNumResolutions.M(1))
+		if ip.IP.To4() != nil {
+			backend = ip.String()
+		} else {
+			// it's an IPv6 address
+			backend = fmt.Sprintf("[%s]", ip.String())
+		}
+
+		for _, port := range r.port {
+			// if a port is specified in the configuration, add it
+			if port != 0 {
+				backends = append(backends, fmt.Sprintf("%s:%d", backend, port))
+			}
+		}
+		return true
+	})
+
+	// keep it always in the same order
+	sort.Strings(backends)
+
+	if equalStringSlice(r.endpoints, backends) {
+		return r.endpoints, nil
+	}
+
+	// the list has changed!
+	r.updateLock.Lock()
+	r.endpoints = backends
+	r.updateLock.Unlock()
+	_ = stats.RecordWithTags(ctx, resolverSuccessTrueMutators, mNumBackends.M(int64(len(backends))))
+
+	// propagate the change
+	r.changeCallbackLock.RLock()
+	for _, callback := range r.onChangeCallbacks {
+		callback(r.endpoints)
+	}
+	r.changeCallbackLock.RUnlock()
+	return r.endpoints, nil
+}
+
+func (r *k8sResolver) onChange(f func([]string)) {
+	r.changeCallbackLock.Lock()
+	defer r.changeCallbackLock.Unlock()
+	r.onChangeCallbacks = append(r.onChangeCallbacks, f)
+}
+
+var _ cache.ResourceEventHandler = (*handler)(nil)
```

Comment:
I would split the cache event handler on its own source file.

Reply:
done